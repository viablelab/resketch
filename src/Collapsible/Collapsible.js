import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const CollapsibleContainer = styled.div`
  border-top: 1px solid silver;

  &:last-of-type {
    border-bottom: 1px solid silver;
  }
`;

const CollapsibleHeader = styled.button`
  background-color: transparent;
  border-style: none;
  border-width: 0;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
  padding: 0.5rem;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  width: 100%;

  &:focus {
    outline-style: none;
    outline-width: 0;
  }
`;

const CollapsibleHeaderTitle = styled.p`
  margin: 0;
`;

const CollapsibleHeaderIcon = styled.div`
  width: 6px;
  height: 6px;
  border-top: 2px solid steelblue;
  border-right: 2px solid steelblue;
  transform: ${props => (!props.isOpen ? 'rotate(45deg)' : 'rotate(135deg)')};
  transition: transform 195ms ease;

  ${CollapsibleHeader}:focus & {
    ${'' /* TODO Add outline on focus */};
  }
`;

const CollapsibleBody = styled.div`
  padding: 0.5rem;
`;

export default class Collapsible extends Component {
  constructor(props) {
    super(props);

    this.toggleOpen = () =>
      this.setState(prevState => ({
        isOpen: !prevState.isOpen,
      }));

    this.state = {
      isOpen: props.initiallyOpen,
    };
  }

  render() {
    const { children, title, ...rest } = this.props;
    const { isOpen } = this.state;

    return (
      <CollapsibleContainer {...rest}>
        <CollapsibleHeader onClick={this.toggleOpen}>
          <CollapsibleHeaderTitle>{title}</CollapsibleHeaderTitle>
          <CollapsibleHeaderIcon isOpen={isOpen} />
        </CollapsibleHeader>
        {isOpen && <CollapsibleBody>{children}</CollapsibleBody>}
      </CollapsibleContainer>
    );
  }
}

CollapsibleContainer.displayName = 'CollapsibleContainer';
CollapsibleHeader.displayName = 'CollapsibleHeader';
CollapsibleHeaderTitle.displayName = 'CollapsibleHeaderTitle';
CollapsibleHeaderIcon.displayName = 'CollapsibleHeaderIcon';
CollapsibleBody.displayName = 'CollapsibleBody';

Collapsible.displayName = 'Collapsible';
Collapsible.propTypes = {
  initiallyOpen: PropTypes.bool,
  title: PropTypes.string.isRequired,
};
Collapsible.defaultProps = {
  initiallyOpen: false,
};
