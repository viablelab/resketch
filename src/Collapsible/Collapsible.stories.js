import React, { Fragment } from 'react';
import { storiesOf } from '@storybook/react';
import Collapsible from './Collapsible';

storiesOf('Collapsible', module)
  .add('single', () => (
    <Collapsible title="What is going on?">
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aut ea hic enim
      veniam aliquid facilis sit, assumenda, iusto soluta dolore explicabo
      voluptatem id nobis ex placeat nihil a ipsum obcaecati.
    </Collapsible>
  ))
  .add('multiple', () => (
    <Fragment>
      <Collapsible title="Item A">
        Lorem ipsum dolor sit amet, consectetur ...
      </Collapsible>
      <Collapsible title="Item B">
        Lorem ipsum dolor sit amet, consectetur ...
      </Collapsible>
      <Collapsible title="Item C">
        Lorem ipsum dolor sit amet, consectetur ...
      </Collapsible>
    </Fragment>
  ));
