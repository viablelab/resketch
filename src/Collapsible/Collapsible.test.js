import React from 'react';
import renderer from 'react-test-renderer';
import Collapsible from './Collapsible';

test('renders properly', () => {
  const rendered = renderer.create(
    <Collapsible title="Title">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Doloremque
        labore neque, ab iste! Necessitatibus debitis alias est quisquam unde
        iste hic cupiditate earum, quis itaque eligendi voluptatem. Ea,
        laudantium, deserunt.
      </p>
    </Collapsible>
  );
  const instance = rendered.root;

  expect(rendered.toJSON()).toMatchSnapshot();

  instance.findByType('button').props.onClick();
  expect(rendered.toJSON()).toMatchSnapshot();
});
