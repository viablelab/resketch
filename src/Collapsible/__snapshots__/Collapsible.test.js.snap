// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
.c0 {
  border-top: 1px solid silver;
}

.c0:last-of-type {
  border-bottom: 1px solid silver;
}

.c1 {
  background-color: transparent;
  border-style: none;
  border-width: 0;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
  padding: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c1:focus {
  outline-style: none;
  outline-width: 0;
}

.c2 {
  margin: 0;
}

.c3 {
  width: 6px;
  height: 6px;
  border-top: 2px solid steelblue;
  border-right: 2px solid steelblue;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  -webkit-transition: -webkit-transform 195ms ease;
  -webkit-transition: transform 195ms ease;
  transition: transform 195ms ease;
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    <p
      className="c2"
    >
      Title
    </p>
    <div
      className="c3"
    />
  </button>
</div>
`;

exports[`renders properly 2`] = `
.c0 {
  border-top: 1px solid silver;
}

.c0:last-of-type {
  border-bottom: 1px solid silver;
}

.c1 {
  background-color: transparent;
  border-style: none;
  border-width: 0;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
  padding: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c1:focus {
  outline-style: none;
  outline-width: 0;
}

.c2 {
  margin: 0;
}

.c3 {
  width: 6px;
  height: 6px;
  border-top: 2px solid steelblue;
  border-right: 2px solid steelblue;
  -webkit-transform: rotate(135deg);
  -ms-transform: rotate(135deg);
  transform: rotate(135deg);
  -webkit-transition: -webkit-transform 195ms ease;
  -webkit-transition: transform 195ms ease;
  transition: transform 195ms ease;
}

.c4 {
  padding: 0.5rem;
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    <p
      className="c2"
    >
      Title
    </p>
    <div
      className="c3"
    />
  </button>
  <div
    className="c4"
  >
    <p>
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Doloremque labore neque, ab iste! Necessitatibus debitis alias est quisquam unde iste hic cupiditate earum, quis itaque eligendi voluptatem. Ea, laudantium, deserunt.
    </p>
  </div>
</div>
`;
